cpp/src/arrow/gpu/cuda_context.cc

```diff
@@ -70,37 +86,41 @@ class CudaContext::CudaContextImpl {
   }
 
   Status CopyHostToDevice(void* dst, const void* src, int64_t nbytes) {
-    CU_RETURN_NOT_OK(cuCtxSetCurrent(context_));
+    ContextSaver set_temporary(context_);
     CU_RETURN_NOT_OK(cuMemcpyHtoD(reinterpret_cast<CUdeviceptr>(dst), src,
                                   static_cast<size_t>(nbytes)));
     return Status::OK();
   }
 
   Status CopyDeviceToHost(void* dst, const void* src, int64_t nbytes) {
-    CU_RETURN_NOT_OK(cuCtxSetCurrent(context_));
+    ContextSaver set_temporary(context_);
     CU_RETURN_NOT_OK(cuMemcpyDtoH(dst, reinterpret_cast<const CUdeviceptr>(src),
                                   static_cast<size_t>(nbytes)));
     return Status::OK();
   }
 
+  Status Synchronize(void) {
+    ContextSaver set_temporary(context_);
+    CU_RETURN_NOT_OK(cuCtxSynchronize());
+    return Status::OK();
+  }
+
   Status Free(void* device_ptr, int64_t nbytes) {
     CU_RETURN_NOT_OK(cuMemFree(reinterpret_cast<CUdeviceptr>(device_ptr)));
     bytes_allocated_ -= nbytes;
     return Status::OK();
   }
 
   Status ExportIpcBuffer(void* data, std::shared_ptr<CudaIpcMemHandle>* handle) {
-    CU_RETURN_NOT_OK(cuCtxSetCurrent(context_));
+    ContextSaver set_temporary(context_);
     CUipcMemHandle cu_handle;
     CU_RETURN_NOT_OK(cuIpcGetMemHandle(&cu_handle, reinterpret_cast<CUdeviceptr>(data)));
     *handle = std::shared_ptr<CudaIpcMemHandle>(new CudaIpcMemHandle(&cu_handle));
     return Status::OK();
   }
 
   Status OpenIpcBuffer(const CudaIpcMemHandle& ipc_handle, uint8_t** out) {
-    CU_RETURN_NOT_OK(cuCtxSetCurrent(context_));
```

Comment:
I'm curious: is ContextSaver not needed here because of the way this API is called in the unit tests? I think we may want to do add it nevertheless, for consistency.

Reply:
No, it is not needed. The unit tests call `CudaContext::OpenIpcBuffer` that uses `ContextSaver`. The referenced OpenIpcBuffer here belongs to `CudaContext::CudaContextImpl` that users or unittests cannot use directly.